File: 977. Squares of a Sorted Array/977.squares-of-a-sorted-array.c
/*
 * @lc app=leetcode id=977 lang=c
 *
 * [977] Squares of a Sorted Array
 */

// @lc code=start
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* sortedSquares(int* nums, int numsSize, int* returnSize) {
    int n = numsSize, left = 0, right = n - 1, position = n - 1;
    int* result = malloc(n * sizeof(int));

    while(left <= right) {
        if(abs(nums[left]) > abs(nums[right])) {
            result[position--] = nums[left] * nums[left];
            left++;
        }
        else {
            result[position--] = nums[right] * nums[right];
            right--;
        }
    }

    *returnSize = n;

    return result;
}
// @lc code=end

File: 977. Squares of a Sorted Array/977.squares-of-a-sorted-array.py
class Solution(object):
    def sortedSquares(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        result = [0] * len(nums) 
        left = 0
        right = len(nums) - 1
        position = len(nums) - 1
        
        while(left <= right):
            if(abs(nums[left]) > abs(nums[right])):
                result[position] = nums[left] ** 2
                left += 1
            else:
                result[position] = nums[right] ** 2
                right -= 1
            position -= 1
        return result